pkg/profiles/tap.go

```diff
@@ -0,0 +1,140 @@
+package profiles
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"sort"
+	"time"
+
+	"github.com/ghodss/yaml"
+	"github.com/linkerd/linkerd2/controller/api/util"
+	sp "github.com/linkerd/linkerd2/controller/gen/apis/serviceprofile/v1alpha1"
+	pb "github.com/linkerd/linkerd2/controller/gen/public"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// RenderTapOutputProfile performs a tap on the desired resource and generates
+// a service profile with routes pre-populated from the tap data
+// Only inbound tap traffic is considered.
+func RenderTapOutputProfile(client pb.ApiClient, tapResource, namespace, name, controlPlaneNamespace string, tapDuration time.Duration, routeLimit int, w io.Writer) error {
+	requestParams := util.TapRequestParams{
+		Resource:  tapResource,
+		Namespace: namespace,
+	}
+	log.Debugf("Running `linkerd tap %s --namespace %s`", tapResource, namespace)
+
+	req, err := util.BuildTapByResourceRequest(requestParams)
+	if err != nil {
+		return err
+	}
+
+	profile, err := tapToServiceProfile(client, req, namespace, name, controlPlaneNamespace, tapDuration, routeLimit)
+	if err != nil {
+		return err
+	}
+
+	output, err := yaml.Marshal(profile)
+	if err != nil {
+		return fmt.Errorf("Error writing Service Profile: %s", err)
+	}
+	w.Write(output)
+	return nil
+}
+
+func tapToServiceProfile(client pb.ApiClient, tapReq *pb.TapByResourceRequest, namespace, name, controlPlaneNamespace string, tapDuration time.Duration, routeLimit int) (sp.ServiceProfile, error) {
+	profile := sp.ServiceProfile{
+		ObjectMeta: meta_v1.ObjectMeta{
+			Name:      fmt.Sprintf("%s.%s.svc.cluster.local", name, namespace),
+			Namespace: controlPlaneNamespace,
+		},
+		TypeMeta: ServiceProfileMeta,
+	}
+
+	ctx, cancel := context.WithDeadline(context.Background(), time.Now().Add(tapDuration))
+	defer cancel()
+
+	tapClient, err := client.TapByResource(ctx, tapReq)
```

Comment:
apologies, my previous recommendation caused a race condition where the tapClient could error out on `tapClient.Recv()` with a scary error message. i think a better approach is to strictly rely on tapClient.Recv() hitting the deadline, and hiding the error if we know it was caused the by deadline. let me know what you think:
```diff
diff --git a/pkg/profiles/tap.go b/pkg/profiles/tap.go
index 01bde4d0..779f12be 100644
--- a/pkg/profiles/tap.go
+++ b/pkg/profiles/tap.go
@@ -6,6 +6,7 @@ import (
        "io"
        "os"
        "sort"
+       "strings"
        "time"

        "github.com/ghodss/yaml"
@@ -62,43 +63,39 @@ func tapToServiceProfile(client pb.ApiClient, tapReq *pb.TapByResourceRequest, n
                return profile, err
        }

-       routes := routeSpecFromTap(ctx, tapClient, routeLimit)
+       routes := routeSpecFromTap(tapClient, routeLimit)

        profile.Spec.Routes = routes

        return profile, nil
 }

-func routeSpecFromTap(ctx context.Context, tapClient pb.Api_TapByResourceClient, routeLimit int) []*sp.RouteSpec {
+func routeSpecFromTap(tapClient pb.Api_TapByResourceClient, routeLimit int) []*sp.RouteSpec {
        routes := make([]*sp.RouteSpec, 0)
        routesMap := make(map[string]*sp.RouteSpec)

-recvLoop:
        for {
-               select {
-               case <-ctx.Done():
-                       break recvLoop
-               default:
-                       log.Debug("Waiting for data...")
-                       event, err := tapClient.Recv()
-
-                       if err == io.EOF {
-                               break recvLoop
-                       }
-
-                       if err != nil {
+               log.Debug("Waiting for data...")
+               event, err := tapClient.Recv()
+
+               if err != nil {
+                       // expected errors when hitting the tapDuration deadline
+                       if err != io.EOF &&
+                               !strings.HasSuffix(err.Error(), context.DeadlineExceeded.Error()) &&
+                               !strings.HasSuffix(err.Error(), "http2: response body closed") {
                                fmt.Fprintln(os.Stderr, err)
-                               break recvLoop
                        }

-                       routeSpec := getPathDataFromTap(event)
-                       log.Debugf("Created route spec: %v", routeSpec)
+                       break
+               }
+
+               routeSpec := getPathDataFromTap(event)
+               log.Debugf("Created route spec: %v", routeSpec)

-                       if routeSpec != nil {
-                               routesMap[routeSpec.Name] = routeSpec
-                               if len(routesMap) > routeLimit {
-                                       break recvLoop
-                               }
+               if routeSpec != nil {
+                       routesMap[routeSpec.Name] = routeSpec
+                       if len(routesMap) > routeLimit {
+                               break
                        }
                }
        }
```

Reply:
oh interesting. yeah I see what you mean. I guess it does simplify the looping if we do is this way!__